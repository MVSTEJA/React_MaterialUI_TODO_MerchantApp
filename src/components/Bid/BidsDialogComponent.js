import React from 'react';
import PropTypes from 'prop-types';
import Table, { TableBody, TableCell, TableHead, TableRow, TableSortLabel } from 'material-ui/Table';
import Typography from 'material-ui/Typography';
import Dialog, { DialogTitle } from 'material-ui/Dialog';

const Constants = {
    BID_ID: 'Bid Id',
    CAR_TITLE: 'Car Title',
    AMOUNT: 'Amount',
    CREATED: 'Created',
    NO_BIDS: 'No Bids'
}
/**
 * @description: A Component for BidsDialog List. 
 */
const BidsDialogComponentList = ({ classes, displayBids, orderBy, order, handleHeaderSort }) => (
    <div className={classes.demo}>
        <Table className={classes.table}>
            <TableHead className={classes.headerStyle}>
                <TableRow>
                    <TableCell className={classes.columnTitle}>{Constants.BID_ID}</TableCell>
                    <TableCell numeric className={classes.columnTitle}>{Constants.CAR_TITLE}</TableCell>
                    <TableCell
                        numeric
                        className={classes.columnTitle}
                        sortDirection={orderBy === 'amount' ? order : false}
                    >
                        <TableSortLabel
                            active={orderBy === 'amount' ? true : false}
                            direction={order}
                            className={classes.columnTitle}
                            onClick={handleHeaderSort.bind(null, 'amount', order)}
                        >
                            {Constants.AMOUNT}
                        </TableSortLabel>
                    </TableCell>
                    <TableCell
                        numeric
                        className={classes.columnTitle}
                        sortDirection={orderBy === 'created' ? order : false}
                    >
                        <TableSortLabel
                            active={orderBy === 'created' ? true : false}
                            direction={order}
                            className={classes.columnTitle}
                            onClick={handleHeaderSort.bind(null, 'created', order)}
                        >
                            {Constants.CREATED}
                        </TableSortLabel>
                    </TableCell>
                </TableRow>
            </TableHead>
            <TableBody>
                {
                    displayBids.map((bid, key) =>
                        <TableRow key={key}>
                            <TableCell
                            >{bid.id}</TableCell>
                            <TableCell
                            >{bid.carTitle}</TableCell>
                            <TableCell
                            >{bid.amount}</TableCell>
                            <TableCell
                            >{bid.created}</TableCell>
                        </TableRow>
                    )
                }
            </TableBody>
        </Table>
    </div >
);

BidsDialogComponentList.propTypes = {
    classes: PropTypes.object.isRequired,
    displayBids: PropTypes.array,
    orderBy: PropTypes.string,
    order: PropTypes.string,
    handleHeaderSort: PropTypes.func
};

export default class BidsDialogComponent extends React.Component {
    state = {
        openBidsDialog: false,
        order: 'asc',
        orderBy: 'amount',
        sortBidLabel: 'amount'
    }

    handleClickBidsOpen = (evt) => {
        evt.preventDefault();
        this.setState({
            openBidsDialog: true
        })
        const { sortBidLabel, order } = this.state;
        this.props.sortBids(this.props.merchant.bids, {
            sortBidLabel,
            order
        });
    }
    handleClickBidsClose = () => {
        this.setState({
            openBidsDialog: false
        })
    }
    handleHeaderSort = (sortBidLabel, order) => {
        const newOrder = order === 'asc' ? 'desc' : 'asc'
        this.props.sortBids(this.props.merchant.bids, {
            sortBidLabel,
            order: newOrder
        });
        this.setState(() => ({
            order: newOrder,
            sortBidLabel,
            orderBy: sortBidLabel
        }))
    }
    render() {
        const { classes, merchant } = this.props;
        if (merchant.bids && merchant.bids.length > 0) {
            return (
                <div>
                    <a href="" className={classes.anchorStyle} onClick={this.handleClickBidsOpen}>
                        <Typography className={classes.anchorStyle}>
                            {`${merchant.bids.length} Bids`}
                        </Typography>
                    </a>
                    <Dialog onClose={this.handleClickBidsClose} open={this.state.openBidsDialog} aria-labelledby="simple-dialog-title">
                        <DialogTitle id="simple-dialog-title">{`Bids by ${merchant.firstName} ${merchant.lastName}`}</DialogTitle>
                        <BidsDialogComponentList
                            classes={classes}
                            displayBids={this.props.displayBids || this.props.merchant.bids}
                            orderBy={this.state.orderBy}
                            order={this.state.order}
                            sortBidLabel={this.props.sortBidLabel || this.state.sortBidLabel}
                            handleHeaderSort={this.handleHeaderSort}
                        />
                    </Dialog>
                </div >
            )
        } else {
            return <div>{Constants.NO_BIDS}</div>
        }

    }
}

BidsDialogComponent.propTypes = {
    classes: PropTypes.object.isRequired,
    sortBids: PropTypes.func,
    displayBids: PropTypes.array,
    sortBidLabel: PropTypes.string,
    merchant: PropTypes.shape({
        bids: PropTypes.array
    })
};