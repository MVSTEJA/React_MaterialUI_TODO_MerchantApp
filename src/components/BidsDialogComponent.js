import React from 'react';

import Table, { TableBody, TableCell, TableHead, TableRow, TableSortLabel } from 'material-ui/Table';

import Typography from 'material-ui/Typography';
import Button from 'material-ui/Button';
import Dialog, { DialogTitle } from 'material-ui/Dialog';

const BidsDialogComponentList = ({ classes, displayBids, orderBy, sortBidLabel, order, handleHeaderSort }) => (
    <div className={classes.demo}>
        <Table className={classes.table}>
            <TableHead className={classes.headerStyle}>
                <TableRow>
                    <TableCell className={classes.columnTitle}>Bid Id</TableCell>
                    <TableCell numeric className={classes.columnTitle}>Car Title</TableCell>
                    <TableCell
                        numeric
                        className={classes.columnTitle}
                        sortDirection={orderBy === 'amount' ? order : false}
                    >
                        <TableSortLabel
                            active={orderBy === 'amount' ? true : false}
                            direction={order}
                            onClick={handleHeaderSort.bind(null, 'amount', order)}
                        >
                            Amount
                        </TableSortLabel>
                    </TableCell>
                    <TableCell
                        numeric
                        className={classes.columnTitle}
                        sortDirection={orderBy === 'created' ? order : false}
                    >
                        <TableSortLabel
                            active={orderBy === 'created' ? true : false}
                            direction={order}
                            onClick={handleHeaderSort.bind(null, 'created', order)}
                        >
                            Created
                        </TableSortLabel>
                    </TableCell>
                </TableRow>
            </TableHead>
            <TableBody>
                {
                    displayBids.map((bid, key) =>
                        <TableRow key={key}>
                            <TableCell
                            >{bid.id}</TableCell>
                            <TableCell
                            >{bid.carTitle}</TableCell>
                            <TableCell
                            >{bid.amount}</TableCell>
                            <TableCell
                            >{bid.created}</TableCell>
                        </TableRow>
                    )
                }
            </TableBody>
        </Table>
    </div >
);

export default class BidsDialogComponent extends React.Component {
    state = {
        openBidsDialog: false,
        order: 'asc',
        orderBy: 'amount',
        sortBidLabel: 'amount'
    }

    handleClickBidsOpen = (evt) => {
        evt.preventDefault();
        this.setState({
            openBidsDialog: true
        })
        const { sortBidLabel, order } = this.state;
        this.props.sortBids(this.props.merchant.bids, {
            sortBidLabel,
            order
        });
    }
    handleClickBidsClose = () => {
        this.setState({
            openBidsDialog: false
        })
    }
    handleHeaderSort = (sortBidLabel, order) => {
        const newOrder = order === 'asc' ? 'desc' : 'asc'
        this.props.sortBids(this.props.merchant.bids, {
            sortBidLabel,
            order: newOrder
        });
        this.setState(prevState => ({
            order: newOrder,
            sortBidLabel,
            orderBy: sortBidLabel
        }))
    }
    render() {
        const { classes, merchant } = this.props;
        if (merchant.bids && merchant.bids.length > 0) {
            return (
                <div>
                    <a href="" className={classes.anchorStyle} onClick={this.handleClickBidsOpen}>
                        <Typography className={classes.anchorStyle}>
                            {`${merchant.bids.length} Bids`}
                        </Typography>
                    </a>
                    <Dialog onClose={this.handleClickBidsClose} open={this.state.openBidsDialog} aria-labelledby="simple-dialog-title">
                        <DialogTitle id="simple-dialog-title">{`Bids by ${merchant.firstName} ${merchant.lastName}`}</DialogTitle>
                        <BidsDialogComponentList
                            classes={classes}
                            displayBids={this.props.displayBids || this.props.merchant.bids}
                            orderBy={this.state.orderBy}
                            order={this.state.order}
                            sortBidLabel={this.props.sortBidLabel || this.state.sortBidLabel}
                            handleHeaderSort={this.handleHeaderSort}
                        />
                    </Dialog>
                </div >
            )
        } else {
            return <div>No Bids</div>
        }

    }
};
