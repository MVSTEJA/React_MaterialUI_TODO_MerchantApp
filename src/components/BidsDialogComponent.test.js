import React from 'react';
import ReactDOM from 'react-dom';
import BidsDialogComponent from './BidsDialogComponent';
import { shallow } from 'enzyme';

import { initialState } from "./../index";

const init = () => {

    return {
        merchant: {},
        classes: {},
        sortBids: jest.fn(),
        event: {
            preventDefault: jest.fn()
        }
    }
}
it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<BidsDialogComponent />, div);
});

jest.mock('react-dom');

it('should shallow mount and match matchsnapshot BidsDialogComponent', () => {
    const { merchant } = init();
    const BidsDialogComponentShallow = shallow(<BidsDialogComponent merchant={merchant} />);

    expect(BidsDialogComponentShallow).toMatchSnapshot();
});

it('should shallow mount with no merchant bids', () => {
    const { merchant, classes } = init();
    merchant.bids = initialState.merchants[0].bids;
    const BidsDialogComponentShallow = shallow(<BidsDialogComponent merchant={merchant} classes={classes} />);

    expect(BidsDialogComponentShallow).toMatchSnapshot();
});

it('should mount with merchant bids', () => {
    const { merchant, classes } = init();
    merchant.bids = initialState.merchants[0].bids;

    const BidsDialogComponentShallow = shallow(
        <BidsDialogComponent merchant={merchant} classes={classes} />
    );

    expect(BidsDialogComponentShallow.find('BidsDialogComponentList').dive()).toMatchSnapshot();
});

it('should invoke handleClickBidsOpen', () => {
    const { merchant, classes, sortBids, event } = init();
    merchant.bids = initialState.merchants[0].bids;

    const BidsDialogComponentShallow = shallow(
        <BidsDialogComponent merchant={merchant} classes={classes} sortBids={sortBids} />
    );

    BidsDialogComponentShallow.instance().handleClickBidsOpen(event);
    expect(BidsDialogComponentShallow.instance().state).toEqual({ "openBidsDialog": true, "order": "asc", "orderBy": "amount", "sortBidLabel": "amount" });
});

it('should invoke handleClickBidsClose', () => {
    const { merchant, classes, sortBids } = init();
    merchant.bids = initialState.merchants[0].bids;

    const BidsDialogComponentShallow = shallow(
        <BidsDialogComponent merchant={merchant} classes={classes} />
    );

    BidsDialogComponentShallow.instance().handleClickBidsClose();
    expect(BidsDialogComponentShallow.instance().state).toEqual({ "openBidsDialog": false, "order": "asc", "orderBy": "amount", "sortBidLabel": "amount" });
});
