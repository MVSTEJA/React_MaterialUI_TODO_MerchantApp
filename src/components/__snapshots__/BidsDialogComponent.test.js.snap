// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match mount with merchant bids 1`] = `
<div>
  <WithStyles(Table)>
    <TableHead
      component="thead"
    >
      <WithStyles(TableRow)>
        <WithStyles(TableCell)>
          Bid Id
        </WithStyles(TableCell)>
        <WithStyles(TableCell)
          numeric={true}
        >
          Car Title
        </WithStyles(TableCell)>
        <WithStyles(TableCell)
          numeric={true}
        >
          Amount
        </WithStyles(TableCell)>
        <WithStyles(TableCell)
          numeric={true}
        >
          Create
        </WithStyles(TableCell)>
      </WithStyles(TableRow)>
    </TableHead>
    <TableBody
      component="tbody"
    >
      <WithStyles(TableRow)
        button={true}
      >
        <WithStyles(TableCell)>
          1
        </WithStyles(TableCell)>
        <WithStyles(TableCell)>
          title
        </WithStyles(TableCell)>
        <WithStyles(TableCell)>
          1
        </WithStyles(TableCell)>
        <WithStyles(TableCell)>
          01/01/01
        </WithStyles(TableCell)>
      </WithStyles(TableRow)>
      <WithStyles(TableRow)
        button={true}
      >
        <WithStyles(TableCell)>
          2
        </WithStyles(TableCell)>
        <WithStyles(TableCell)>
          title2
        </WithStyles(TableCell)>
        <WithStyles(TableCell)>
          2
        </WithStyles(TableCell)>
        <WithStyles(TableCell)>
          02/02/02
        </WithStyles(TableCell)>
      </WithStyles(TableRow)>
    </TableBody>
  </WithStyles(Table)>
</div>
`;

exports[`should match shallow mount and matchsnapshot BidsDialogComponent 1`] = `
<div>
  No Bids
</div>
`;

exports[`should match shallow mount with no merchant bids 1`] = `
<div>
  <WithStyles(Button)
    dense={true}
    onClick={[Function]}
    raised={true}
  >
    Bids
  </WithStyles(Button)>
  <WithStyles(Dialog)
    aria-labelledby="simple-dialog-title"
    onClose={[Function]}
    open={false}
  >
    <WithStyles(DialogTitle)
      id="simple-dialog-title"
    >
      Bids
    </WithStyles(DialogTitle)>
    <BidsDialogComponentList
      displayBids={
        Array [
          Object {
            "amount": 1,
            "carTitle": "title",
            "created": "01/01/01",
            "id": "1",
          },
          Object {
            "amount": 2,
            "carTitle": "title2",
            "created": "02/02/02",
            "id": "2",
          },
        ]
      }
    />
  </WithStyles(Dialog)>
</div>
`;
